lib/crypto.js

```diff
@@ -661,6 +661,33 @@ function filterDuplicates(names) {
   }).sort();
 }
 
+// This implements Brad Hill's Double HMAC pattern from
+// https://www.nccgroup.trust/us/about-us/
+//   newsroom-and-events/blog/2011/february/double-hmac-verification/.
+// In short, it's near-impossible to write a reliable constant-time compare in a
+// high level language like JS, because of the many layers that can optimize
+// away attempts at being constant time.
+//
+// Double HMAC avoids that problem by blinding the timing channel instead. After
+// running the inputs through a second round of HMAC, we are free to
+// short-circuit comparison, because the time it takes to reach the
+// short-circuit has no relation to the similarity between the guessed digest
+// and the correct one.
+exports.timingSafeEqual = timingSafeEqual;
+function timingSafeEqual(a, b) {
+  if (!(a instanceof Buffer))
+    throw new TypeError('First argument must be a Buffer');
+  if (!(b instanceof Buffer))
+    throw new TypeError('Second argument must be a Buffer');
+  var key = randomBytes(32);
+  var ah = new Hmac('sha256', key).update(a).digest();
+  var bh = new Hmac('sha256', key).update(b).digest();
+  // The second test, for a.equals(b), is just in case of the vanishingly small
+  // chance of a collision. It only fires if the digest comparison passes and so
+  // doesn't leak timing information.
+  return ah.equals(bh) && a.equals(b);
```

Comment:
```Remember that we are comparing that risk versus the current status quo, which is that it is impossible to safely validate an HMAC with the current library.```

Sorry to be so harsh, but in the interest of user safety, I think it would actually be far safer for users to do whatever they are currently doing, than use a function which does `a.equals(b)`.